### ISSUE TITLE
Have Accounting Connection Sync Spinner Timeout After 20 minutes.

### ISSUE THREAD

**Issue by yuwenmemon** [Expensify/App#42356]
# Part of the #Wave-Control NetSuite project
Main issue: https://github.com/Expensify/Expensify/issues/377671
Doc section: https://docs.google.com/document/d/1WubNv_VAv78IxG4FKsi9aS0pWESfWvUYbqBAAy5b4bc/edit#heading=h.tb9yro1dmj3r
Project: https://github.com/orgs/Expensify/projects/130

# Feature Description
One issue with the Pusher ProgressManager updates for syncs that come from the IS to NewDot is that they are sometimes unreliable. If the [`jobDone`](https://github.com/Expensify/App/blob/947f9c686f23ad6fbeff10b07d62db32e4161ecf/src/CONST.ts#L1796) sync update is not received by the client, then the sync spinner will hang forever, even though the sync has finished:
 

https://github.com/Expensify/App/assets/4741899/1360d0e9-cab0-49b6-a302-24a88f477e03

To fix this, let's have the spinner timeout to stop 20 min after a sync is started and no `jobDone` sync update is received. We landed one 20 minutes because, in [recent measurements](https://expensify.slack.com/archives/C06ML6X0W9L/p1715981896075939?thread_ts=1715817436.742279&cid=C06ML6X0W9L), our longest NetSuite jobs take around 14 minutes.

# Manual Test Steps
1. Initiate a QBO sync from your workspace that's connected to QBO
2. Go offline
3. Wait for 10 min
4. Go back online
5. Make sure that after 10 more minutes, the spinner has stopped.

# Automated Tests
N/A

cc @aldo-expensify @francoisl @lakchote @marcaaron @tgolen

**Comment by twisterdotcom** [comment]
I think this is `New Feature`, not `Bug` right?

**Comment by yuwenmemon** [comment]
Yeah that seems like a fair characterization as well.

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Spinner hasn't stopped even though the sync has finished

### What is the root cause of that problem?
This is a new feature

### What changes do you think we should make in order to solve the problem?
1. In `QuickBooksOnline`, create a function like `syncConnectionAfterTimeout` that will update `stageInProgress` to `jobDone`


2. In `PolicyAccountingPage`, we will create a `useEffect` that will be triggered when `connectionSyncProgress?.stageInProgress` is changed and in this we will have a setTimeout 20 minutes to update `connectionSyncProgress?.stageInProgress` to `jobDone` so if a sync is started and no `jobDone` sync update is received

```
useEffect(() => {
    if (!connectionSyncProgress?.stageInProgress || connectionSyncProgress?.stageInProgress === CONST.POLICY.CONNECTIONS.SYNC_STAGE_NAME.JOB_DONE) {
        return;
    }
    spinnerTimeoutRef.current = setTimeout(() => {
        syncConnectionAfterTimeout(policy.id);
    }, 20*60*1000);

    return () => {
        if (!spinnerTimeoutRef.current) {
            return;
        }

        clearTimeout(spinnerTimeoutRef.current)
    }
}, [connectionSyncProgress?.stageInProgress])
```


### What alternative solutions did you explore? (Optional)
We can create a state to control whenever we should show spinner and then update it in the `setTimeout` of point 2

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Spinner hasn't stopped even though the sync has finished

### What is the root cause of that problem?
This is a new feature

### What changes do you think we should make in order to solve the problem?
1. In `QuickBooksOnline`, create a function like `syncConnectionAfterTimeout` that will update `stageInProgress` to `jobDone`

```
function syncConnectionAfterTimeout(policyID: string) {
    Onyx.merge(`${ONYXKEYS.COLLECTION.POLICY_CONNECTION_SYNC_PROGRESS}${policyID}`,  {
        stageInProgress: CONST.POLICY.CONNECTIONS.SYNC_STAGE_NAME.JOB_DONE,
        connectionName: CONST.POLICY.CONNECTIONS.NAME.QBO,
    })
}
```
2. In `PolicyAccountingPage`, we will create a `useEffect` that will be triggered when `connectionSyncProgress?.stageInProgress` is changed and in this we will have a setTimeout 20 minutes to update `connectionSyncProgress?.stageInProgress` to `jobDone` so if a sync is started and no `jobDone` sync update is received

```
useEffect(() => {
    if (!connectionSyncProgress?.stageInProgress || connectionSyncProgress?.stageInProgress === CONST.POLICY.CONNECTIONS.SYNC_STAGE_NAME.JOB_DONE) {
        return;
    }
    spinnerTimeoutRef.current = setTimeout(() => {
        syncConnectionAfterTimeout(policy.id);
    }, 20*60*1000);

    return () => {
        if (!spinnerTimeoutRef.current) {
            return;
        }

        clearTimeout(spinnerTimeoutRef.current)
    }
}, [connectionSyncProgress?.stageInProgress])
```


### What alternative solutions did you explore? (Optional)
We can create a state to control whenever we should show spinner and then update it in the `setTimeout` of point 2

**Comment by aldo-expensify** [comment]
@yuwenmemon for this, instead of adding a `setTimeout` in the front end, I think it can be more reliably implemented like:


1. The IS server should include a timestamp in the progress data ([PolicyConnectionSyncProgress](https://github.com/Expensify/App/blob/e8ae3c5acedf0e6788dc574c7b6f3043ca37092a/src/types/onyx/Policy.ts#L545))
2. [isSyncInProgress](https://github.com/Expensify/App/blob/e8ae3c5acedf0e6788dc574c7b6f3043ca37092a/src/pages/workspace/accounting/PolicyAccountingPage.tsx#L115) should ignore the progress data if it is older than 20 minutes

What do you think?

**Comment by francoisl** [comment]
Does that solve the issue with missing the final notification though? Let's say you close your laptop while a sync is running, and so you never effectively receive the `jobDone` notification. When you get back online, you still have the wrong `stageInProgress` right?

**Comment by aldo-expensify** [comment]
> When you get back online, you still have the wrong stageInProgress right?

yes, it would be still in Onyx, but ignored because it is older than 20 minutes

**Comment by yuwenmemon** [comment]
@aldo-expensify Yeah good idea, that should make things a lot simpler and reliable. In this case we'll need to take this internal.

**Comment by francoisl** [comment]
Ah ok I originally misunderstood your comment, I thought you wanted to ignore new progress data received by IS if the previous one was older than 20 minutes. Makes more sense now 👍

**Comment by twisterdotcom** [comment]
Removed the /app job and automation hopefully.

**Comment by twisterdotcom** [comment]
Ah crap, wait I see @ahmedGaber93 reviewed https://github.com/Expensify/App/pull/42498. Damn, thought that was an internal PR. Okay, will pay that out when it's due.

**Comment by twisterdotcom** [comment]
Payment Summary:

- @ahmedGaber93 paid $250 [here](https://www.upwork.com/nx/wm/offer/102808540) (C+)

**Comment by ahmedGaber93** [comment]
@twisterdotcom offer accepted, please please complete the payment when you are available. Thanks!
